File: backend/src/services/product-sort.util.ts
import type { ProductSort } from "../types/product.types.js";

export function buildProductSort(sort: ProductSort): Record<string, 1 | -1> {
  if (sort === "title-desc") {
    return { name: -1, slug: 1 };
  }

  if (sort === "price-desc") {
    return { "variants.0.price": -1, slug: 1 };
  }

  if (sort === "price-asc") {
    return { "variants.0.price": 1, slug: 1 };
  }

  return { name: 1, slug: 1 };
}

File: backend-springboot/src/main/java/com/season/app/util/ProductSortUtil.java
package com.season.app.util;

import org.springframework.data.domain.Sort;

public final class ProductSortUtil {
    private ProductSortUtil() {
    }

    public static Sort buildProductSort(String sort) {
        if ("title-desc".equals(sort)) {
            return Sort.by(Sort.Order.desc("name"), Sort.Order.asc("slug"));
        }

        if ("price-desc".equals(sort)) {
            return Sort.by(Sort.Order.desc("variants.0.price"), Sort.Order.asc("slug"));
        }

        if ("price-asc".equals(sort)) {
            return Sort.by(Sort.Order.asc("variants.0.price"), Sort.Order.asc("slug"));
        }

        return Sort.by(Sort.Order.asc("name"), Sort.Order.asc("slug"));
    }
}

